python/ray/train/tests/test_examples.py

```diff
@@ -19,50 +20,49 @@
     train_func as fashion_mnist_train_func,
 )
 from ray.train.examples.train_linear_example import train_func as linear_train_func
+from ray.train.horovod.horovod_trainer import HorovodTrainer
+from ray.train.tensorflow.tensorflow_trainer import TensorflowTrainer
 from ray.train.tests.test_trainer import KillCallback
+from ray.train.torch.torch_trainer import TorchTrainer
 
 
 @pytest.fixture
-def ray_start_2_cpus():
-    address_info = ray.init(num_cpus=2)
+def ray_start_4_cpus():
+    address_info = ray.init(num_cpus=4)
     yield address_info
     # The code after the yield will run as teardown code.
     ray.shutdown()
 
 
 @pytest.mark.parametrize("num_workers", [1, 2])
-def test_tensorflow_mnist(ray_start_2_cpus, num_workers):
+def test_tensorflow_mnist(ray_start_4_cpus, num_workers):
     num_workers = num_workers
     epochs = 3
 
-    trainer = Trainer("tensorflow", num_workers=num_workers)
     config = {"lr": 1e-3, "batch_size": 64, "epochs": epochs}
-    trainer.start()
-    results = trainer.run(tensorflow_mnist_train_func, config)
-    trainer.shutdown()
+    trainer = TensorflowTrainer(
+        tensorflow_mnist_train_func,
+        train_loop_config=config,
+        scaling_config=dict(num_workers=num_workers),
+    )
+    results = trainer.fit()
 
-    assert len(results) == num_workers
-    result = results[0]
+    result = results.metrics
 
-    loss = result["loss"]
-    assert len(loss) == epochs
-    assert loss[-1] < loss[0]
-
-    accuracy = result["accuracy"]
-    assert len(accuracy) == epochs
-    assert accuracy[-1] > accuracy[0]
```

Comment:
Can we keep these assert statements? We should still test that accuracy improves right?

Reply:
We only have access to the last result in the `Result` object :/

Reply:
I am not sure what's the best approach here. I do agree we should test for this. I can think of either creating a debug trainer that will return the trial dataframe, or we should revamp Trainer (and possibly Tune) so that the entire Trial history is returned in a `Result`.

Reply:
Changes in https://github.com/ray-project/ray/pull/25943 will allow to test this again

Reply:
added back in